include/openthread/tcp.h

```diff
@@ -0,0 +1,521 @@
+ * empties the send and receive buffer, transferring ownership of any data
+ * provided by the application in @p otTcpActiveSendByReference calls back to
+ * the application. The socket's callbacks and memory for the receive buffer
+ * remain associated with the socket.
+ *
+ * @param[in]  aInstance  A pointer to the OpenThread instance to which the socket belongs.
+ * @param[in]  aSocket    A pointer to the active TCP socket structure representing the TCP endpoint to abort.
+ *
+ * @retval OT_ERROR_NONE    Successfully closed the active TCP socket.
+ * @retval OT_ERROR_FAILED  Failed to close the active TCP socket.
+ */
+otError otTcpActiveAbort(otInstance *aInstance, otTcpActiveSocket *aSocket);
+
+/**
+ * Closes this active TCP socket. If it corresponds to a live TCP connection,
+ * the connection is terminated unceremoniously (as in @p otTcpActiveAbort).
+ * All resources the application has provided for this socket (linked buffers
+ * for the send buffer, memory for the receive buffer, the @p otTcpActiveSocket
+ * structure itself, etc.) are immediately returned to the application. This is
+ * different from the classic BSD API, in which the TCP connection may outlive
+ * the socket.
+ *
+ * @param[in]  aInstance  A pointer to the OpenThread instance to which the socket belongs.
+ * @param[in]  aSocket    A pointer to the active TCP socket structure to close.
+ *
+ * @retval OT_ERROR_NONE    Successfully closed the active TCP socket.
+ * @retval OT_ERROR_FAILED  Failed to close the active TCP socket.
+ */
+otError otTcpActiveClose(otInstance *aInstance, otTcpActiveSocket *aSocket);
+
+/**
+ * This callback indicates that an incoming connection that matches this
+ * passive socket has arrived. The application should provide (via the
+ * callback's return value) a pointer to the @p otTcpActiveSocket into which
+ * to accept the incoming connection. If the application returns OT_TCP_DEFER,
+ * then OpenThread silently ignores the connection establishment request; the
+ * callback will be called again the next time that the connection
+ * establishment request is submitted. This is valuable if resources are not
+ * available to accept the connection, but they may be available at the time of
+ * a future attempt. If the application returns OT_TCP_REFUSE, then OpenThread
+ * sends a "connection refused" message to the host that attempted to establish
+ * a connection.
+ */
+typedef otTcpActiveSocket *(*otTcpAcceptReady)(void *aContext, const otSockAddr *aPeer);
+
+/**
+ * This callback indicates that the TCP connection is now ready for two-way
+ * communication. In the case of TCP Fast Open, this may be before the TCP
+ * connection handshake has actually completed. The application is provided
+ * with the context pointers both for the passive socket that accepted the
+ * connection and the active socket into which it was accepted.
+ */
+typedef void (*otTcpAcceptDone)(void *aPassiveContext, void *aActiveContext, const otSockAddr *aPeer);
```

Comment:
What are `aPassiveContext` and `aActiveContext`? 
Should we include a pointer to the socket `otTcpPassiveSockt *`?

Reply:
`aPassiveContext` is the context associated with the passive socket that accepted a connection. `aActiveContext` is the context associated with the active socket into which the connection was accepted. The context was associated with each socket at the time it was created.

As for replacing this with a pointer to the socket directly, see my comment above about using the context pointer rather than a pointer to the socket to conform with OpenThread's existing APIs. I'm open to discussing this further. :)

Reply:
See [here](https://github.com/openthread/openthread/pull/6491#discussion_r617101245) for a discussion of this. I included this in my latest update, according to your suggestion. Let's continue the conversation [here](https://github.com/openthread/openthread/pull/6491#discussion_r617101245) so that everyone's thoughts are in one place. Meanwhile, I'll resolve this conversation.